### PR TITLE
Updates to E(H)CalEnergy definitions for PF hadrons.

### PR DESCRIPTION
**Pull request by hatakeyamak**
#### PR description:

Based on recent discussion in the PF meeting, the following changes are being implemented:
+ Improve rawE(H)CalEnergy definitions for neutral hadrons.
   - Neutral hadrons (photons) from "loop hcal" are not associated with ECAL (HCAL) clusters, as they are defined based on excess energies in HCAL (ECAL) energies only. rawE(H)CalEnergy is now set to zero for neutral hadrons (photons), as done for corrected energies to avoid confusions.
+ protect against the major source of negative ecal energy for charged hadrons.
   - Occasionally ecal energy for charged hadrons went negative, when photons are also created and the ecal energy corrected by PF hadron calibration is smaller than ecal energy from egamma's ecal cluster calibration. Now this is protected against going negative.
+ also protect against negative energy scaling.
   - Very rarely, charged hadron momentum scaling went negative, after rejecting of "bad" tracks. Now we protect against it as well. 

https://indico.cern.ch/event/832017/contributions/3485635/attachments/1871517/3102604/PF_NeutralHadron_CaloFractions_v4.pdf
for further information.

#### PR validation:

It compiles. Also, some validation plots in the above pdf file.

#### if this PR is a backport please specify the original PR:

This is not a backport.

@bendavid @zdemirag @ahinzmann

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by fabiocos** [comment]
@slava77 @mrodozov the recent update of workflow names 2023 -> 2026 in #27488 requires to update the map in the bot for comparisons

**Comment by fabiocos** [comment]
see https://github.com/cms-sw/cms-bot/pull/1170

**Comment by slava77** [comment]
@cmsbuild please test

to refresh with now appropriately configured reco comparisons in the cms-bot

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc [hunk 1]
```diff
@@ -2311,8 +2315,9 @@ void PFAlgo::createCandidatesHCAL(const reco::PFBlock& block,
           particleDirection.push_back(std::get<1>(pae));
           ecalEnergy.push_back(mergedPhotonEnergy * clusterEnergyCalibrated / sumEcalClusters);
           hcalEnergy.push_back(0.);
-          rawecalEnergy.push_back(totalEcal);
-          rawhcalEnergy.push_back(totalHcal);
+          rawecalEnergy.push_back(mergedPhotonEnergy * clusterEnergyCalibrated /
+                                  sumEcalClusters);  // "raw" not well defined. use the corrected value.
+          rawhcalEnergy.push_back(0.);
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/src/PFAlgo.cc]: this breaks the HO energy settings .

Some protections are necessary below in 
https://github.com/cms-sw/cmssw/blob/18425f38932959335ace0f686b663eacfa129e69/RecoParticleFlow/PFProducer/src/PFAlgo.cc#L2374-L2377

Comment (hatakeyamak) [reply RecoParticleFlow/PFProducer/src/PFAlgo.cc]: Thank you @slava77 . I made the change.

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/27550	0458365
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show differences primarily in the PF candidate energy fractions, as expected
    - in AOD/RECO: tau emFraction is now non-negative (ignoring default "-1" values), which propagates to various tau discriminants as small changes
    - in miniAOD caloFraction and hcalFraction of PF candidates change in sync and hcalFraction is now bound in [0,1] range
    - also, in some rate cases (only in wf 25202 among all the tested wfs, including  jenkins and also some limited particle guns, dijets, and ttbar with PU and 2018 data in local tests) there is a change in PF charged momentum

**Comment by kpedro88** [comment]
+1

**Pull request merged**